### PR TITLE
[torchdim] Fix Python 3.11 bytecode decoding in dims

### LINKED ISSUE #91246: First class dims doesn't work with python 3.11
Trying to do `a, b = dims()` will always raise `SyntaxError: dims() must be assigned to a sequence of variable names or have argument n specified`.

As far as I can tell, the OPCODE detection at https://github.com/pytorch/pytorch/blob/15af4b1ceea3b689354ad664675b930486804c8e/functorch/csrc/dim/dim.cpp#LL1531C7-L1531C7 is not working anymore.

cc @zou3519 @Chillee @samdow @soumith @zdevito

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91290

Adds a PyInstDecoder object that handles the differences in bytecode
added in 3.11. Basically some instructions have inline caches which
change the size of the instruction, so calculating the next instruction
is slightly different.

fixes #91246

### CODE REVIEW

**Comment by albanD** [comment]
Dont hesitate to use https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/python_compat.h and https://github.com/pytorch/pytorch/blob/master/torch/csrc/utils/pythoncapi_compat.h to reduce the need to ifdef. The functions there also handle the change in behavior of some functions from one version to the next (like lasti).
These are pure C headers so will be safe to include anywhere.

**Comment by zdevito** [comment]
Updated to remove the ifdefs possible. Looks like it passes on old pythons on the CI, and I verified locally it works on 3.11 (CPU).

File: functorch/csrc/dim/dim.cpp [hunk 1]
```diff
@@ -21,6 +21,11 @@
 #include "arena.h"
 #include "python_variable_simple.h"
 
+#if PY_VERSION_HEX >= 0x030b0000
```
Comment (albanD) [inline functorch/csrc/dim/dim.cpp:24]: nit: You have a `IS_PYTHON_3_11_PLUS`

File: functorch/csrc/dim/dim.cpp [hunk 2]
```diff
@@ -1491,6 +1469,50 @@ py::object create_dimlist(py::object name, py::handle size) {
...
+                names = py::object::borrow(code_object_->co_names);
+                break;
+            case STORE_FAST:
+                names = py::object::steal(PyCode_GetVarnames(code_object_));
+                break;
+            case STORE_DEREF:
+                names = py::object::steal(PyCode_GetCellvars(code_object_));
+                break;
+            default:
+                return py::object();
```
Comment (albanD) [inline functorch/csrc/dim/dim.cpp:1506]: Do you actually want to return an empty object?
```suggestion
                Py_RETURN_NONE

**Review by albanD — APPROVED** [review]
Sounds good!

**Pull request closed without merging**